File: 0212. Word Search II.java
// Hard
// Trie, DFS, Backtracking
// Time:O(k·L + m·n·4^L),Space:O(k·L)
// https://leetcode.cn/problems/word-search-ii/

import java.util.*;

class Solution {
    int[][] DIRS = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};


    public List<String> findWords(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String w : words) {
            trie.insert(w);
        }

        Set<String> res = new HashSet<String>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs(board, trie, i, j, res);
            }
        }
        return new ArrayList<String>(res);
    }

    private void dfs(char[][] board, Trie cur, int i, int j, Set<String> res) {
        if (!cur.child.containsKey(board[i][j])) return;

        char c = board[i][j];
        cur = cur.child.get(c); // get the cur char's trie node 
        if (!"".equals(cur.word)) {
            res.add(cur.word);
        }

        board[i][j] = '#';
        for (int[] dir : DIRS) {
            int i2 = i + dir[0], j2 = j + dir[1];
            if (i2 >= 0 && i2 < board.length && j2 >= 0 && j2 < board[0].length) {
                dfs(board, cur, i2, j2, res);
            }
        }
        board[i][j] = c;
    }
}

class Trie {
    // Time:O(m*n),Space:O(m*n)
    String word;
    Map<Character, Trie> child;

    public Trie() {
        this.word = "";
        this.child = new HashMap<Character, Trie>();
    }

    public void insert(String word) {
        Trie cur = this;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (!cur.child.containsKey(c)) {
                cur.child.put(c, new Trie());
            }
            cur = cur.child.get(c);
        }
        cur.word = word;
    }
}

File: 0001-0500/0212. Word Search II.py
# Time:O(m*n*4^L), Space:O(W*L)

from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        for word in words:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.word = word

        m, n = len(board), len(board[0])
        res = []

        def dfs(i, j, node):
            char = board[i][j]
            if char not in node.children:
                return
            next_node = node.children[char]

            if next_node.word:
                res.append(next_node.word)
                next_node.word = None

            board[i][j] = "#"

            for di, dj in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and board[ni][nj] != "#":
                    dfs(ni, nj, next_node)

            board[i][j] = char

        for i in range(m):
            for j in range(n):
                dfs(i, j, root)

        return res
